R/rc_initialise.R

```diff
@@ -1,38 +1,363 @@
+    }
+  
+  
+  # Define rothc parameters
+  k1 <- rothc.parms$k1
+  k2 <- rothc.parms$k2
+  k3 <- rothc.parms$k3
+  k4 <- rothc.parms$k4
+  
+  abc <- rothc.parms$abc
+  
+  # initialise options
+  
+  ## do a simulation for 150 years to estimate the C fractions assuming system is in equilibrium
+  if(initialization_method =='spinup_simulation'){
+   
+   # Extend crop and amendment files to include entire 150 year simulation
+    crop_extend <- if(!is.null(crops)){
+      rc_extend_crops(start_date = start_date, simyears = 150, crops = crops)
+    }else{
+      NULL
+    }
+    
+    amendment_extend <- if(!is.null(amendment)){
+      rc_extend_amendments(start_date = start_date, simyears = 150, amendments = amendment)
+    }else{
+      NULL
+    }
+ 
+    
+    # Set model parameters
+    parms <- list(unit = 'psomperfraction',
+                  initialization_method = 'none')
+    
+    # Set newly required inputs
+  
+    # Run initialization run for 150 years
+    this.result <- rc_sim(rothc_rotation = crop_extend,
+                          rothc_amendment = amendment_extend,
+                          soil_properties = soil_properties,
+                          weather = dt.weather,
+                          rothc_parms = parms)
+    
+    # take last two rotations
+    this.result.fin <- this.result[year > max(year)-2*nrow(crops),lapply(.SD,mean)]
   
-  # Prepare input for scenario Business As Usual
-  scen.inp <- rc_input_scenario(B_LU_BRP = B_LU_BRP, scen = 'BAU')
-  rotation <- scen.inp$rotation
-  amendment <- scen.inp$amendment
   
-  # Set model parameters
-  parms <- list(simyears = 150,unit = 'psomperfraction', initialize = TRUE)
+    fractions <- this.result.fin[,.(fr_IOM = CIOM / A_SOM_LOI,
+                                    fr_DPM = CDPM / A_SOM_LOI,
+                                    fr_RPM = CRPM / A_SOM_LOI,
+                                    fr_BIO = CBIO / A_SOM_LOI)]
+    
+  }
   
-  # Run initialization run for 150 years
-  this.result <- rc_sim(soil_properties = soil_properties,
-                               rothc_rotation = rotation,
-                               rothc_amendment = amendment,
-                               rothc_parms = parms)
+  ## calculate initial carbon pools assuming equilibrium in C pools, using analytical solution (Heuvelink)
+  if(initialization_method == 'spinup_analytical_heuvelink'){
+
+    # establish simyear based on time data table
+    isimyears <- max(dt.time$year) - min(dt.time$year) + 1
+    
+    # averaged total C input (kg C/ha/year) from crops and amendments
+    if (is.null(crops) || nrow(crops) == 0) {
+      c_input_crop <- 0
+    }else{
+    c_input_crop <- crops[,sum(B_C_OF_INPUT)/isimyears]
+    }
+    
+    if (is.null(amendment) || nrow(amendment) == 0) {
+       c_input_man <- 0
+    } else if (!is.null(amendment$B_C_OF_INPUT) && any(!is.na(amendment$B_C_OF_INPUT))) {
+        c_input_man <- amendment[, sum(B_C_OF_INPUT)/isimyears]
```

Comment:
is this now a difficult way to check that "B_C_OF_INPUT" %in% colnames(amendment)?
why adding the second check, since you are also not doing that for P_DOSE below (you add na.rm=T).
so, it can be c_input_man <- amendment[, sum(B_C_OF_INPUT, na.rm=TRUE)/isimyears]